### PR TITLE
chore(insights): A cached variant for every query response

### PR DESCRIPTION
**Pull request by Twixes**
~~## Problem~~

~~Currently `HogQLQuery` and `EventsQuery` aren't cached in any way, which is pretty bad in terms of loading dashboards. If there is a specific reason for this, absolutely let me know! But my perspective currently is that it's overzealous to calculate new results every single time. We already have `_is_stale()` for determining staleness.~~

~~## Changes~~

~~All query runner-based queries now use the query runner caching layer.~~

See comment below restating this PRs goals.

### CODE REVIEW

**Review by mariusandra — APPROVED** [review]

**Comment by Twixes** [comment]
@mariusandra I see you've re-ran the backend tests, but they aren't doing that badly, they've caught some actual issues. 😅 The big one is that HogQLQueryResponse doesn't conform to CachedQueryResponse, because the former has some extra unexpected fields, and CachedQueryResponse isn't prepared to deal with this.

I would prefer if CachedQueryResponse only had a field like `child: HogQLQueryResponse | TrendsQueryResponse | etc.` which would allow us to deal better with variability, but we'll have to work around the current schema somehow. Possibly just adding five more fields to CachedQueryResponse, which would unfortunately be unused for almost all other queries

**Comment by mariusandra** [comment]
Hmm, whoops, my bad. Saw a lot of failures, yet my bad connection in the last cafe didn't let me look deeper. I could at least load the "run again" button, so that's what I clicked. The first test I clicked on did have a 2min runtime before flaking as well.

Regarding the five extra fields, what about just one complex object with a key like `cacheControl` or such? Nothing really against the extra fields, as we do already throw a lot of metadata and custom fields into all `DataNode`-s (e.g. `hogql` in the output, `modifiers` in the input, etc)

**Comment by Twixes** [comment]
Unfortunately this seemingly simple quest required shaving a yak, and this is a pretty big yak as far as yaks go. And they don't go very far, because they weight between 200 and 700 kg.

## Problem

Our Pydantic models for the cores of query responses are great. Each query type has its own shape of a response. When something is an `EventsQueryResponse`, you can sleep well knowing it has the right attributes matching the expected types.

But not cached responses! For these we only had `CachedQueryResponse`. It was clearly designed for classic insight queries, whose responses are pretty barebones. But let's say we want to cache an SQL insight – its response has a few extra, SQL-specific response fields.

Now, we could turn `CachedQueryResponse` into a monster with lots of optional fields, but then all hell can break loose with what's inside…

## Changes

…or we can ensure that every `*QueryResponse` has its well-formed `Cached*QueryResponse` equivalent. Thanks to the power of TypeScript generics and `extends`, we can.

File: mypy-baseline.txt
Comment (Twixes) [inline mypy-baseline.txt:1]: The typing improvements eliminate a bunch of mypy errors. No errors added.

File: frontend/src/scenes/debug/QueryTabs.tsx [hunk 1]
```diff
@@ -33,13 +33,13 @@ function toLine(hogql: string, position: number): number {
...
     return toLineColumn(hogql, position).column
 }
-interface QueryTabsProps {
-    query: Node
+interface QueryTabsProps<Q extends Node> {
+    query: Q
     queryKey: `new-${string}`
-    response?: Record<string, any> | null
-    setQuery: (query: DataNode) => void
+    response?: Q['response'] | null
```
Comment (Twixes) [inline frontend/src/scenes/debug/QueryTabs.tsx:39]: There's a bunch of changes like this. Thanks to better guarantees, specifically on `response` here, TypeScript was complaining before the switch to generics here. Generics are definitely the accurate way here though.

File: frontend/src/queries/schema.ts
Comment (Twixes) [inline frontend/src/queries/schema.ts:1]: This is the core of the changes.

File: posthog/hogql_queries/query_runner.py
Comment (Twixes) [inline posthog/hogql_queries/query_runner.py:1]: Query runner base updated here

**Review by mariusandra — APPROVED** [review]
That was one hairy yak. The changes look great!

**Pull request merged**